File: lib/PluginUpgrade/example/upgradeExample.py
import struct
import time
import zlib

import numpy as np
import serial
from typing import NamedTuple


def padarray(A, size):
    t = size - len(A)
    return np.pad(A, pad_width=(0, t), mode='constant')


class InfoAction(NamedTuple):
    hardwareVersion: int
    flashStart: int
    programVersion: int
    programCrc: int


class SealHeader(NamedTuple):
    deviceId: int
    version: int
    size: int
    crc: int


class Seal(NamedTuple):
    hardwareVersion: SealHeader
    programStart: int
    programSize: int
    programCrc: int
    programVersion: int


def checkError(line):
    if line == b"false \r\n":
        print("Error something happened")
        exit(255)
    pass


def send(data):
    print(data,end='')
    print()
    ser.write(data)

def flash():
    # Lest get the info
    send(b"upgrade.keyscanner.getInfo\n")
    line = ser.readline()
    ser.readline()
    checkError(line)
    line = line.replace(b" \r\n", b"")
    line = line.split(b" ")
    info: InfoAction = InfoAction(
        int(line[0]),
        int(line[1]),
        int(line[2]),
        int(line[3]),
    )
    print("This is the info")
    print(info)
    # Here we can check the version and crc
    program = open("KeyScanner_WithHeader.bin", mode='rb').read()
    # The first 32 bits are the seal.
    sealInBinary = struct.unpack(b"<IIIIIIII", program[0:32])
    sealInBinary = Seal(
        SealHeader(sealInBinary[0], sealInBinary[1], sealInBinary[2], sealInBinary[3]),
        sealInBinary[4], sealInBinary[5], sealInBinary[6], sealInBinary[7])
    print(sealInBinary)
    if sealInBinary.programCrc == info.programCrc:
        print("No need to update!")
        # exit(0)
    for i in range(0, len(program), 256):
        print("Writing " + str(i) + " of " + str(len(program)))
        writeAction = struct.pack(b"<II", i + info.flashStart, 256)
        data = program[i:(i + 256)]
        crc32Transmission = struct.pack(b"<I", zlib.crc32(data))
        send(str.encode("upgrade.keyscanner.sendWrite ")+writeAction+data+crc32Transmission)
        line = ser.readline()
        checkError(line)
        ser.readline()

    send(b"upgrade.keyscanner.finish\n")
    ser.readline()
    ser.readline()


def main():
    global ser
    # configure the serial connections (the parameters differs on the device you are connecting to)
    ser = serial.Serial(
        port='/dev/ttyACM1',
        baudrate=115200,
    )
    ser.isOpen()
    send(b"upgrade.start\n")
    ser.readline()
    ser.readline()
    while True:
        send(b"upgrade.isReady\n")
        line1 = ser.readline()

        ser.readline()

        if line1 == b"true \r\n":
            break
        time.sleep(1)

    print("Can start upgrading")

    send(b"upgrade.keyscanner.beginLeft\n")
    line = ser.readline()
    checkError(line)
    ser.readline()
    flash()

    send(b"upgrade.keyscanner.beginRight\n")
    line = ser.readline()
    checkError(line)
    ser.readline()
    flash()


main()

File: lib/PluginUpgrade/example/upgradeExample.js
const fs = require("fs");
const {crc32} = require("easy-crc");
const {SerialPort} = require("serialport");
const {ReadlineParser} = require("@serialport/parser-readline");


async function init() {
    const hexFile = fs.readFileSync('KeyScanner_WithHeader.bin', "hex")
    const fromHexString = (hexString) => Uint8Array.from(hexString.match(/.{1,2}/g).map((byte) => parseInt(byte, 16)));
    const binaryFile = fromHexString(hexFile);
    const serialport = new SerialPort({path: '/dev/ttyACM1', baudRate: 115200})
    const parser = serialport.pipe(new ReadlineParser({ delimiter: '\r\n' }))
    let receivedData = [];
    parser.on('data', function (data) {
        receivedData.push(data)
    })


    async function readLine() {
        while (receivedData.length === 0) await sleep(1)
        return receivedData.pop();
    }



    serialport.write("upgrade.keyscanner.start\n")
    console.log(await readLine())
    console.log("You have 4 seconds to press the button!")
    await sleep(4000);
    serialport.write("upgrade.keyscanner.beginLeft\n")
    console.log(await readLine())
    console.log(await readLine())

    serialport.write("upgrade.keyscanner.getInfo\n")
    console.log(await readLine())
    console.log(await readLine())

    let info = {
        hardwareVersion: 1,
        flashStart: 20480,
        programVersion: 16777217,
        programCrc: 3782824883
    }
    for (let i = 0; i < binaryFile.length; i=i+256) {
        serialport.write("upgrade.keyscanner.sendWrite ")
        const writeAction = new Uint8Array(new Uint32Array([info.flashStart+i,256]).buffer)
        const data = binaryFile.slice(i,i+256);
        const crc = new Uint8Array(new Uint32Array([crc32("CRC-32",data)]).buffer)
        const blob =  new Uint8Array(writeAction.length+data.length+crc.length)
        blob.set(writeAction)
        blob.set(data,writeAction.length)
        blob.set(crc,data.length+writeAction.length)
        const buffer = new Buffer.from(blob);
        serialport.write(buffer)
        console.log(await readLine())
        console.log(await readLine())
    }

    serialport.write("upgrade.keyscanner.finish\n")
    console.log(await readLine())
    console.log(await readLine())

    process.exit(1);
}

function sleep(ms) {
    return new Promise((resolve) => {
        setTimeout(resolve, ms);
    });
}


init()